src/main/java/com/ably/kafka/connect/utils/RecordHeaderConversions.java

```diff
@@ -0,0 +1,136 @@
+package com.ably.kafka.connect.utils;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import com.google.gson.internal.LinkedTreeMap;
+import io.ably.lib.types.MessageExtras;
+import io.ably.lib.util.JsonUtils;
+import org.apache.kafka.connect.header.Header;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.util.Base64;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class RecordHeaderConversions {
+
+    private static final Logger logger = LoggerFactory.getLogger(RecordHeaderConversions.class);
+    private static final String PUSH_HEADER = "com.ably.extras.push";
+    private static final String KAFKA_KEY = "kafka";
+    private static final String PUSH_KEY = "push";
+    /**
+     * Returns the extras object to use when converting a Kafka message
+     * to an Ably message.
+     * <p>
+     * If the Kafka message has a key, it is base64 encoded and set as the
+     * "key" field in the extras and is added under "kafka" extras.
+     * <p>
+     * If the Kafka message has headers, they are set as the "headers" field
+     * in the extras.
+     *
+     * If the Kafka message has "com.ably.extras.push" header this is set as "push" extras for Ably message.
+     *
+     * @param record The sink record representing the Kafka message
+     * @return The Kafka message extras object
+     */
+    @Nullable
+    public static MessageExtras toMessageExtras(final SinkRecord record) {
+        JsonUtils.JsonUtilsObject kafkaExtras = null;
+        JsonUtils.JsonUtilsObject extrasObject = null;
+
+        Object pushExtras = null;
+        final byte[] key = (byte[]) record.key();
+        if (key != null) {
+            kafkaExtras = JsonUtils.object();
+            final String keyValue = Base64.getEncoder().encodeToString(key);
+            kafkaExtras.add("key", keyValue);
+        }
+
+        if (record.headers().isEmpty()){
+            if (kafkaExtras == null) {
+                return null;
+            }
+            extrasObject = JsonUtils.object();
+            extrasObject.add(KAFKA_KEY, kafkaExtras);
+            return new MessageExtras(extrasObject.toJson());
+        }
+
+        JsonUtils.JsonUtilsObject headersObject = null;
+        for (Header header : record.headers()) {
+            if (header.key().equals(PUSH_HEADER)) {
+                pushExtras = header.value();
+            } else {
+                if (kafkaExtras == null) {
+                    kafkaExtras = JsonUtils.object();
+                }
+                if (headersObject == null) {
+                    headersObject = JsonUtils.object();
+                }
+                headersObject.add(header.key(), header.value());
+            }
+        }
+
+        if (kafkaExtras == null && pushExtras == null) {
```

Comment:
I'm wondering - we have a lot of places that have to check whether `kafkaExtras` and `headersObject` etc are null, and instantiate them if not, which could make the code harder to follow in the future (and potentially more prone to NullPointerExceptions).

Should we just instantiate the objects straight up, and then just do `.size() == 0` to check for items? It may make things a bit easier to follow

Reply:
I wasn't particularly happy with this long method so I ended up writing a wrapper object and builder for it to make it easier to follow the logic. 1918c37cf4bf6bb6c23d26869153db9b0ba24eba